oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java

```diff
@@ -0,0 +1,296 @@
+        this.elasticConnection = elasticConnection;
+        this.indexDefinition = indexDefinition;
+        this.definitionBuilder = definitionBuilder;
+        this.bulkProcessor = initBulkProcessor();
+    }
+
+    /**
+     * Returns an ElasticBulkProcessorHandler instance based on the index definition configuration.
+     * <p>
+     * The `sync-mode` property can be set to `rt` (real-time). In this case the returned handler will be real-time.
+     * This option is available for sync index definitions only.
+     */
+    public static ElasticBulkProcessorHandler getBulkProcessorHandler(@NotNull ElasticConnection elasticConnection,
+                                                                      @NotNull ElasticIndexDefinition indexDefinition,
+                                                                      @NotNull NodeBuilder definitionBuilder, CommitInfo commitInfo) {
+        PropertyState async = indexDefinition.getDefinitionNodeState().getProperty("async");
+
+        if (async != null) {
+            return new ElasticBulkProcessorHandler(elasticConnection, indexDefinition, definitionBuilder);
+        }
+
+        // commit-info has priority over configuration in index definition
+        String syncMode = null;
+        if (commitInfo != null) {
+            syncMode = (String) commitInfo.getInfo().get(SYNC_MODE_PROPERTY);
+        }
+
+        if (syncMode == null) {
+            PropertyState syncModeProp = indexDefinition.getDefinitionNodeState().getProperty("sync-mode");
+            if (syncModeProp != null) {
+                syncMode = syncModeProp.getValue(Type.STRING);
+            }
+        }
+
+        if (SYNC_RT_MODE.equals(syncMode)) {
+            return new RealTimeBulkProcessorHandler(elasticConnection, indexDefinition, definitionBuilder);
+        }
+
+        return new ElasticBulkProcessorHandler(elasticConnection, indexDefinition, definitionBuilder);
+    }
+
+    private BulkProcessor initBulkProcessor() {
+        return BulkProcessor.builder(requestConsumer(),
+                new OakBulkProcessorListener())
+                .setBulkActions(indexDefinition.bulkActions)
+                .setBulkSize(new ByteSizeValue(indexDefinition.bulkSizeBytes))
+                .setFlushInterval(TimeValue.timeValueMillis(indexDefinition.bulkFlushIntervalMs))
+                .setBackoffPolicy(BackoffPolicy.exponentialBackoff(
+                        TimeValue.timeValueMillis(indexDefinition.bulkRetriesBackoff), indexDefinition.bulkRetries)
+                )
+                .build();
+    }
+
+    protected BiConsumer<BulkRequest, ActionListener<BulkResponse>> requestConsumer() {
+        return (request, bulkListener) -> elasticConnection.getClient().bulkAsync(request, RequestOptions.DEFAULT, bulkListener);
+    }
+
+    public void add(DocWriteRequest<?> request) {
+        bulkProcessor.add(request);
+        totalOperations++;
+    }
+
+    public boolean close() {
+        if (totalOperations <= 0) {
```

Comment:
Nit: I would use "totalOperations != 0", because < 0 isn't possible (well in theory it could become negative). But it's a matter of taste of course.